src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/security/GithubAuthenticator.kt

```diff
@@ -1,46 +1,117 @@
 package org.vaccineimpact.orderlyweb.security
 
-import org.pac4j.core.context.Pac4jConstants
+import org.eclipse.egit.github.core.User
+import org.eclipse.egit.github.core.client.GitHubClient
+import org.eclipse.egit.github.core.client.RequestException
+import org.eclipse.egit.github.core.service.OrganizationService
+import org.eclipse.egit.github.core.service.TeamService
+import org.eclipse.egit.github.core.service.UserService
 import org.pac4j.core.context.WebContext
 import org.pac4j.core.credentials.TokenCredentials
 import org.pac4j.core.credentials.authenticator.Authenticator
 import org.pac4j.core.exception.CredentialsException
 import org.pac4j.core.profile.CommonProfile
 import org.pac4j.core.util.CommonHelper
+import org.vaccineimpact.orderlyweb.db.AppConfig
+import org.vaccineimpact.orderlyweb.db.Config
+import org.vaccineimpact.orderlyweb.db.UserData
+import org.vaccineimpact.orderlyweb.errors.BadConfigurationError
 
-class GithubAuthenticator : Authenticator<TokenCredentials>
+
+class GithubAuthenticator(private val userData: UserData,
+                          private val githubApiClient: GitHubClient,
+                          private val appConfig: Config = AppConfig()) : Authenticator<TokenCredentials>
 {
     override fun validate(credentials: TokenCredentials?, context: WebContext)
     {
         if (credentials == null)
         {
-            throwsException("No credentials supplied")
+            throw CredentialsException("No credentials supplied")
+        }
+
+        val token = credentials.token
+
+        if (CommonHelper.isBlank(token))
+        {
+            throw CredentialsException("Token cannot be blank")
+        }
+
+        val email = validate(token)
+        credentials.userProfile = CommonProfile().apply {
+            this.addAttribute("url", "*")
+            this.setId(email)
+        }
+
+    }
+
+    private fun validate(token: String): String
+    {
+        githubApiClient.setOAuth2Token(token)
+
+        val user = getGitHubUser()
+
+        val githubOrg = appConfig["app.github_org"]
+        val teamName = appConfig["app.github_team"]
+
+        if (!githubOrg.isEmpty() && !userBelongToOrg(githubOrg, user))
+        {
+            throw CredentialsException("User is not a member of GitHub org $githubOrg")
         }
-        else
+        if (!githubOrg.isEmpty() && !teamName.isEmpty() && !userBelongsToTeam(githubOrg, teamName, user))
         {
-            val token = credentials.token
+            throw CredentialsException("User is not a member of GitHub team $teamName")
+        }
```

Comment:
The logic here means that if no org is specified in the config any GitHub user can log in. Not sure whether this is desired behaviour, or whether to throw an error if no org is specified...